File: src/telegram/senderName.ts
import { Api, type TelegramClient } from "telegram";

async function resolveUserName(client: TelegramClient, peer: Api.PeerUser): Promise<string> {
	try {
		const user = await client.getEntity(peer);
		if (user instanceof Api.User) {
			if (user.username) {
				return `@${user.username}`;
			}
			if (user.firstName || user.lastName) {
				return [user.firstName, user.lastName]
					.filter(Boolean)
					.join(" ")
					.trim();
			}
			return `User ${user.id}`;
		}
	} catch (error) {
		console.error("Error getting sender display name:", error);
	}
	return "Unknown";
}

export async function getSenderDisplayName(
	client: TelegramClient,
	message: Api.Message,
): Promise<string> {
	if (message.fromId instanceof Api.PeerUser) {
		return resolveUserName(client, message.fromId);
	}

	// In private chats fromId is often null — the sender is the peer itself
	if (!message.fromId && message.peerId instanceof Api.PeerUser) {
		return resolveUserName(client, message.peerId);
	}

	return "Unknown";
}

File: src_py/telegram_utils/sender_name.py
import logging

from telethon import TelegramClient
from telethon.tl import types

logger = logging.getLogger(__name__)


async def _resolve_user_name(client: TelegramClient, peer: types.PeerUser) -> str:
    try:
        user = await client.get_entity(peer)
        if isinstance(user, types.User):
            if user.username:
                return f"@{user.username}"
            parts = [user.first_name, user.last_name]
            name = " ".join(p for p in parts if p).strip()
            if name:
                return name
            return f"User {user.id}"
    except Exception:
        logger.exception("Error getting sender display name")
    return "Unknown"


async def get_sender_display_name(
    client: TelegramClient, message: types.Message
) -> str:
    if isinstance(message.from_id, types.PeerUser):
        return await _resolve_user_name(client, message.from_id)

    if not message.from_id and isinstance(message.peer_id, types.PeerUser):
        return await _resolve_user_name(client, message.peer_id)

    return "Unknown"
